File: Palindrome Partitioning.py
class Solution:
	def isPalindrome(self, str):
		i = 0
		j = len(str) - 1
		while (i < j):
			if (str[i] != str[j]):
				return False
			i = i + 1
			j = j - 1
		return True
	def doPartition(self, str, substrs, result):
		for i in range(1, len(str)+1):
			substr = str[0:i]
			if (not self.isPalindrome(substr)):
				continue
			substrs.append(substr)
			if (i == len(str)):
				result.append(copy.copy(substrs))
			else:
				self.doPartition(str[i:], substrs, result)
			substrs.pop()
	
	# @param s, a string
	# @return a list of lists of string
	def partition(self, s):
		substrs = []
		result = []
		self.doPartition(s, substrs, result)
		return result

File: Palindrome Partitioning.java
public class Solution {
    private boolean isPalindrome(String s) {
        int i = 0;
        int j = s.length() - 1;
        while (i < j) {
            if (s.charAt(i) != s.charAt(j))
                return false;
            i++;
            j--;
        }
        return true;
    }
    private void partition(String s, List<String> substrs, List<List<String>> result) {
        for (int len = 1; len <= s.length(); len++) {
            String substr = s.substring(0, len);
            if (!isPalindrome(substr))
                continue;
            substrs.add(substr);
            if (len == s.length()) {
                List<String> sol = new ArrayList<String>(substrs);
                result.add(sol);
            }
            partition(s.substring(len), substrs, result);
            substrs.remove(substrs.size() - 1);
        }
    }
    public List<List<String>> partition(String s) {
        List<List<String>> result = new ArrayList<List<String>>();
        List<String> substrs = new ArrayList<String>();
        partition(s, substrs, result);
        return result;
    }
}